core/src/main/scala/org/apache/spark/scheduler/AsyncEventQueue.scala

```diff
@@ -154,8 +154,9 @@ private class AsyncEventQueue(
       return
     }
 
-    eventCount.incrementAndGet()
-    if (eventQueue.offer(event)) {
+    if (eventQueue.offer(event, conf.get(LISTENER_BUS_EVENT_QUEUE_TIMEOUT),
+      TimeUnit.MILLISECONDS)) {
```

Comment:
We cannot block here - too many codepaths, including in scheduler, will get blocked.

Reply:
Blocking only occurs when the eventQueue is full, and a full eventQueue means that the load on the current driver is already relatively high, and continuing to allow individual events to be generated will affect the overall speed of processing jobs. After several tests on my end, when the queue is full and the timeout is set to 30s, there is instead a performance improvement of around 5% in a high concurrency scenario (compared to before the modification)

Reply:
While minimizing event drop is helpful, the event subsystem has been written to ensure it continues to work with lost events and prioritize stability of internal subsystems by remaining non blocking. Blocking here is not an option.

For specific applications exhibiting the issue, you can increase event queue size and driver memory- but this is unfortunately not a solution.